extensions/interactions/TextInput/directives/text-input-validation.service.ts

```diff
@@ -111,41 +111,66 @@ export class TextInputValidationService {
     let seenStringsStartsWith: string[] = [];
     let seenStringsEquals: string[] = [];
     let seenStringsFuzzyEquals: string[] = [];
+    let seenRuleTypes: Set<string>;
 
     for (let [answerGroupIndex, answerGroup] of answerGroups.entries()) {
+      seenRuleTypes = new Set();
       for (let [ruleIndex, rule] of answerGroup.rules.entries()) {
-        let currentString = <string> rule.inputs.x;
+        if (seenRuleTypes.has(rule.type)) {
+          warningsList.push({
+            type: AppConstants.WARNING_TYPES.ERROR,
+            message: (
+              `Answer group ${answerGroupIndex + 1} has multiple rules with ` +
+              `the same type \'${rule.type}\' within the same group.`
+            )
+          });
+        }
```

Comment:
Isn't this really likely to trigger failures in existing explorations? The migration above seems to only be migrating the type of inputs, not de-duplicating the rule types.

Reply:
I believe the migration does de-duplicate within each AnswerGroup.